ledger/src/slot_stats.rs

```diff
@@ -103,30 +100,32 @@ impl SlotsStats {
         source: ShredSource,
         slot_meta: Option<&SlotMeta>,
     ) {
-        let mut slot_full_reporting_info = None;
-        let mut stats = self.stats.lock().unwrap();
-        let (slot_stats, evicted) = Self::get_or_default_with_eviction_check(&mut stats, slot);
-        match source {
-            ShredSource::Recovered => slot_stats.num_recovered += 1,
-            ShredSource::Repaired => slot_stats.num_repaired += 1,
-            ShredSource::Turbine => {
-                *slot_stats
-                    .turbine_fec_set_index_counts
-                    .entry(fec_set_index)
-                    .or_default() += 1
+        let (slot_full_reporting_info, evicted) = {
+            let mut stats = self.stats.lock().unwrap();
+            let (slot_stats, evicted) = Self::get_or_default_with_eviction_check(&mut stats, slot);
+            match source {
+                ShredSource::Recovered => slot_stats.num_recovered += 1,
+                ShredSource::Repaired => slot_stats.num_repaired += 1,
+                ShredSource::Turbine => {
+                    *slot_stats
+                        .turbine_fec_set_index_counts
+                        .entry(fec_set_index)
+                        .or_default() += 1
+                }
             }
-        }
-        if let Some(meta) = slot_meta {
-            if meta.is_full() {
-                slot_stats.last_index = meta.last_index.unwrap_or_default();
-                if !slot_stats.flags.contains(SlotFlags::FULL) {
-                    slot_stats.flags |= SlotFlags::FULL;
-                    slot_full_reporting_info =
-                        Some((slot_stats.num_repaired, slot_stats.num_recovered));
+            let mut slot_full_reporting_info = None;
+            if let Some(meta) = slot_meta {
+                if meta.is_full() {
+                    slot_stats.last_index = meta.last_index.unwrap();
```

Comment:
This looks safe to me. Namely, `is_full()` being true implies `last_index.is_some()` per the last line of this function
https://github.com/anza-xyz/agave/blob/0b4657adb2cf7cba68b3018347af456b285e67ef/ledger/src/blockstore_meta.rs#L576-L601

I do agree with you that some future cleanup could be nice to avoid the `.unwrap()` but we can defer that